ProjFS.Mac/PrjFSKext/PrjFSKext/VirtualizationRoots.cpp

```diff
@@ -352,50 +352,60 @@ VirtualizationRootResult VirtualizationRoot_RegisterProviderForPath(PrjFSProvide
+            err = vn_getpath(virtualizationRootVNode, virtualizationRootCanonicalPath, &virtualizationRootCanonicalPathLength);
             
-            RWLock_AcquireExclusive(s_virtualizationRootsLock);
+            if (0 == err)
             {
-                rootIndex = FindRootAtVnode_Locked(virtualizationRootVNode, rootVid, vnodeIds);
-                if (rootIndex >= 0)
+                FsidInode vnodeIds = Vnode_GetFsidAndInode(virtualizationRootVNode, vfsContext);
+                uint32_t rootVid = vnode_vid(virtualizationRootVNode);
+                
+                RWLock_AcquireExclusive(s_virtualizationRootsLock);
                 {
-                    // Reattaching to existing root
-                    if (nullptr != s_virtualizationRoots[rootIndex].providerUserClient)
+                    rootIndex = FindRootAtVnode_Locked(virtualizationRootVNode, rootVid, vnodeIds);
+                    if (rootIndex >= 0)
                     {
-                        // Only one provider per root
-                        err = EBUSY;
-                        rootIndex = RootHandle_None;
+                        // Reattaching to existing root
+                        if (nullptr != s_virtualizationRoots[rootIndex].providerUserClient)
+                        {
+                            // Only one provider per root
+                            err = EBUSY;
+                            rootIndex = RootHandle_None;
+                        }
+                        else
+                        {
+                            VirtualizationRoot& root = s_virtualizationRoots[rootIndex];
+                            root.providerUserClient = userClient;
+                            root.providerPid = clientPID;
+                            virtualizationRootVNode = NULLVP; // transfer ownership
+                        }
                     }
                     else
                     {
-                        VirtualizationRoot& root = s_virtualizationRoots[rootIndex];
-                        root.providerUserClient = userClient;
-                        root.providerPid = clientPID;
-                        virtualizationRootVNode = NULLVP; // transfer ownership
+                        rootIndex = InsertVirtualizationRoot_Locked(userClient, clientPID, virtualizationRootVNode, rootVid, vnodeIds, virtualizationRootCanonicalPath);
+                        if (rootIndex >= 0)
+                        {
+                            assert(rootIndex < s_maxVirtualizationRoots);
+                        
+                            virtualizationRootVNode = NULLVP; // prevent vnode_put later; active provider should hold vnode reference
+                        
+                            KextLog_Note("VirtualizationRoot_RegisterProviderForPath: new root not found in offline roots, inserted as new root with index %d. path '%s'", rootIndex, virtualizationRootPath);
```

Comment:
In the log message use `virtualizationRootCanonicalPath` rather than `virtualizationRootPath` (or include both if you think we'd potentially care about them being different).

Reply:
Fixed, thanks!